### PR TITLE
Boundary Condition Reader interface

### PR DESCRIPTION
**Pull request by juliasloan25**
For reviewers: This is a large PR, but the majority of the changes are in `src/BCReader.jl`, `test/bcreader_tests.jl`, and `test/mpi_tests/bcreader_mpi_tests.jl`. Also, I created the `src/CallbackManager.jl` file in this PR to simplify compatibility with BCReader, but that module will be documented and tested in a separate PR.

Move functions from `coupler_utils/bcfile_reader.jl` to `src/BCReader.jl` module, and ensure that they are thoroughly commented and tested.

#134 

## To-do
PHASE 1 - moving and documentation
- [x] Copy file(s) to module in `src/`
- [x] Add module to `ClimaCoupler.jl`
- [x] Docstring for each function
  - [x] Add input/output types in docstring if not type annotated
  - [x] Add docstring info to `docs/src`, `docs/make.jl`
- [x] Add exports for externally-used functions
- [x] Flag potentially-removable parts of code

PHASE 2 - testing
- [x] Create test file in `test/`, add to `test/runtests.jl`
- [x] Unit test for each exported function in module
- [x] Integration tests as needed
- [x] Add any tests that use MPI in `test/mpi_tests` and in buildkite pipeline

PHASE 3 - cleanup
- [x] Remove flagged parts of code
- [x] Check all imports are necessary
- [x] Formatting
- [x] Read through comments

### CODE REVIEW

File: src/BCReader.jl [hunk 1]
```diff
@@ -0,0 +1,310 @@
+#=
+    BCReader
+
+This module coordinates reading of boundary conditions from NetCDF files,
+as well as regridding calls and temporal interpolations from
+monthly to daily intervals.
+=#
+module BCReader
+
+using ClimaCoupler: ClimaCoupler, Utilities, Regridder, CallbackManager
```
Comment (charleskawczynski) [inline src/BCReader.jl:10]:
```suggestion
using ..: Utilities, Regridder, CallbackManager
```
This is probably safer (`using ClimaCoupler` can result in using multiple versions of the package, and can cause very confusing errors)

File: src/BCReader.jl [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+- `date`: [Dates.DateTime] start date for data.
+- `bcf_info`: [BCFileInfo] containing boundary condition data.
+"""
+function update_midmonth_data!(date, bcf_info::BCFileInfo)
+    # monthly count
+    all_dates = bcf_info.all_dates
+    midmonth_idx = bcf_info.segment_idx[1]
+    midmonth_idx0 = bcf_info.segment_idx0[1]
+    outfile_root = bcf_info.hd_outfile_root
+    scaling_function = bcf_info.scaling_function
```
Comment (charleskawczynski) [inline src/BCReader.jl:184]:
```suggestion
    (; hd_outfile_root, scaling_function) = bcf_info
```
Note that unpacking can be done like this, too

File: src/CallbackManager.jl [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#=
+    CallbackManager - TODO WIP
+
+This module facilitates calendar functions and temporal interpolations
+of data.
+=#
+
+module CallbackManager
+
+using ClimaCoupler: ClimaCoupler, Utilities
```
Comment (charleskawczynski) [inline src/CallbackManager.jl:10]:
```suggestion
using ..: Utilities
```

File: test/bcreader_tests.jl [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            sst_data = AW.get_data_folder(
+                AW.ArtifactWrapper(
+                    @__DIR__,
+                    "sst",
+                    AW.ArtifactFile[AW.ArtifactFile(
+                        url = "https://caltech.box.com/shared/static/8gd3wjq2dbrnzuv8pd5ww3a54h0kkcz8.nc",
+                        filename = "sst.nc",
+                    ),],
+                ),
+            )
```
Comment (charleskawczynski) [inline test/bcreader_tests.jl:217]: Note that this can lead to a race condition (if this same code is exercised in different buildkite jobs), since artifacts are saved to a common path. Please see [this PR](https://github.com/CliMA/ClimaAtmos.jl/pull/981) for an example of how this race condition can be avoided.

File: test/bcreader_tests.jl [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+            # construct weightfile name to test values
+            hd_outfile_root = varname * "_cgll"
+            outfile = hd_outfile_root * ".nc"
+            outfile_root = mono ? outfile[1:(end - 3)] * "_mono" : outfile[1:(end - 3)]
+            weightfile = joinpath(regrid_dir, outfile_root * "_remap_weights.nc")
+            weights = NCDataset(weightfile)["S"]
+
+            # test monotone remapping (all weights in [0, 1])
+            max_weight = maximum(weights)
+            min_weight = minimum(weights)
```
Comment (charleskawczynski) [inline test/bcreader_tests.jl:300]:
```suggestion
           nt = NCDataset(weightfile)["S"] do weights

                # test monotone remapping (all weights in [0, 1])
                max_weight = maximum(weights)
                min_weight = minimum(weights)
                (; max_weight, min_weight)
            end
            (; max_weight, min_weight) = nt
```

`NCDataset(filename::AbstractString)` opens a file. I recommend using the `do` syntax to ensure that it's always closed, otherwise it's a memory leak

File: test/bcreader_tests.jl [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            monthly_fields,                     # monthly_fields
+            segment_length,                     # segment_length
+            nothing,                            # scaling_function
+            interpolate_daily,                  # interpolate_daily
+            nothing,                            # land_mask
+        )
+        @test BCReader.interpolate_midmonth_to_daily(FT, date0, bcf_info_no_interp) == monthly_fields[1]
+    end
+
+    # Add tests which use TempestRemap here -
```
Comment (LenkaNovak) [inline test/bcreader_tests.jl:118]: Don't the tests below use TR?

File: test/bcreader_tests.jl [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+            topology = Topologies.Topology2D(mesh)
+            quad = Spaces.Quadratures.GLL{Nq}()
+            boundary_space_t = Spaces.SpectralElementSpace2D(topology, quad)
+            current_fields = Fields.zeros(FT, boundary_space_t), Fields.zeros(FT, boundary_space_t)
+
+            land_mask_t = Fields.zeros(boundary_space_t)
+            dummy_data = (; test_data = zeros(axes(land_mask_t)))
+
+            datafile_rll = sst_data
+            varname = "SST"
```
Comment (LenkaNovak) [inline test/bcreader_tests.jl:142]: is this using the original file with all of the SST dates? This would make the test quite slow. I wonder if we use a more lightweight version (e.g. save only a couple of the dates in a separate file in the Box as [here](https://github.com/CliMA/ClimaCoupler.jl/wiki/ClimaCoupler-Lessons-Learned#Working-with-external-data-files)).

**Review by LenkaNovak — COMMENTED** [review]
Looks good! I had a couple of comments and agree with Charlie's suggestions. Also, it would be good to implement this in the `coupler_driver_modular.jl` file once the #167 is merged.

File: test/bcreader_tests.jl (hunk 7 above)
Comment (juliasloan25) [reply test/bcreader_tests.jl]: This is a great idea. Do you know how we can select a subset of the dates in the file when or after we download it?

File: test/bcreader_tests.jl (hunk 6 above)
Comment (juliasloan25) [reply test/bcreader_tests.jl]: The two tests below this both call `bcfile_info_init`, which uses TR so it can't be run on windows

File: test/bcreader_tests.jl (hunk 5 above)
Comment (charleskawczynski) [reply test/bcreader_tests.jl:301]: cc @juliasloan25

**Review by charleskawczynski — COMMENTED** [review]
This is a pretty big PR, could it maybe be broken into smaller parts?

Also, I suggest we start using SafeTestsets.jl in the test suite to avoid leaky tests-- this can cause some really unintuitive errors

File: src/BCReader.jl [hunk 8]
```diff
@@ -0,0 +1,309 @@
+#=
+    BCReader
+
+This module coordinates reading of boundary conditions from NetCDF files,
+as well as regridding calls and temporal interpolations from
+monthly to daily intervals.
+=#
```
Comment (charleskawczynski) [inline src/BCReader.jl:7]: Just FYI, this can be a doc string (`#= this is a multi-line comment =#`

File: src/BCReader.jl [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    varname::V
+    all_dates::D
+    segment_idx::Vector{Int}
+    segment_idx0::Vector{Int}
+    monthly_fields::C
+    segment_length::Vector{Int}
+    scaling_function::O
+    interpolate_daily::Bool
+    land_mask::M
+end
```
Comment (charleskawczynski) [inline src/BCReader.jl:53]:
```suggestion
struct BCFileInfo{FT, B, X, S, V, D, C, O, M}
    bcfile_dir::B
    comms_ctx::X
    hd_outfile_root::S
    varname::V
    all_dates::D
    segment_idx::Vector{Int}
    segment_idx0::Vector{Int}
    monthly_fields::C
    segment_length::Vector{Int}
    scaling_function::O
    interpolate_daily::Bool
    land_mask::M
end
float_type(::BCFileInfo{FT}) where {FT} = FT
```
Storing types in struct fields is actually a performance issue-- the `typeof(Float64)` is a `DataType`, and the compiler cannot optimize fields of this type.

By chance, was this pattern adopted from somewhere else?

Comment (juliasloan25) [reply src/BCReader.jl:53]: It was - I'll open a little PR to address all the places in the code this style is used

Comment (juliasloan25) [reply src/BCReader.jl:53]: It seems like with the new style, no default constructors are created for the struct. I created a MWE 
`struct Foo{FT, S}
       val::S
       end`, which then gives `methods(Foo) 0 methods for type constructor`.
Is there an easy way around this? I could write explicit constructors for all the structs, but they're pretty large.

File: src/BCReader.jl [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+export BCFileInfo, bcfile_info_init, update_midmonth_data!, next_date_in_file, interpolate_midmonth_to_daily
+
+
+"""
+    BCFileInfo
+
+Stores information specific to each boundary condition from a file and each variable.
+
+# Inputs
+    FT::F                           # float type
```
Comment (valeriabarra) [inline src/BCReader.jl:25]: If you use Charlie's suggestion, remember to update the docstring :)

File: src/CallbackManager.jl [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+=#
+
+module CallbackManager
+
+using ..Utilities
+using Dates
+
+export current_date, strdate_to_datetime, datetime_to_strdate, calendar_callback
+
+# TODO remove calls to calendar_callback - can be replaced with `<`
```
Comment (valeriabarra) [inline src/CallbackManager.jl:15]: Is this TODO for a later PR or the comment is outdated? I don't see calls to `calendar_callback` in this file

Comment (juliasloan25) [reply src/CallbackManager.jl:15]: This `TODO` is for a later PR when I go through the `CallbackManager` module in more detail. I just put the reminder here since it's where `calendar_callback` is defined.

File: src/BCReader.jl (hunk 10 above)
Comment (juliasloan25) [reply src/BCReader.jl:25]: Will do!

File: src/BCReader.jl (hunk 9 above)
Comment (charleskawczynski) [reply src/BCReader.jl:53]: There are a few options, one is to make a generic-like outer constructor:
```
julia> struct Foo{FT, S} val::S end

julia> Foo{FT}(args...) where {FT} = Foo{FT, typeof.(args)...}(args...)

julia> Foo{Float64}(1)
Foo{Float64, Int64}(1)
```
Another option is basically that with the use of the `Base.@kwdef` for adding a keyword constructor

File: docs/src/bcreader.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    ClimaCoupler.BCReader.interpolate_midmonth_to_daily
+```
+
+
+## BCReader Internal Functions
+
+```@docs
+    ClimaCoupler.BCReader.no_scaling
+    ClimaCoupler.BCReader.interpol
+```
```
Comment (charleskawczynski) [inline docs/src/bcreader.md:23]:
```suggestion
```@docs
ClimaCoupler.BCReader.BCFileInfo
ClimaCoupler.BCReader.bcfile_info_init
ClimaCoupler.BCReader.update_midmonth_data!
ClimaCoupler.BCReader.next_date_in_file
ClimaCoupler.BCReader.interpolate_midmonth_to_daily
```


## BCReader Internal Functions

```@docs
ClimaCoupler.BCReader.no_scaling
ClimaCoupler.BCReader.interpol
```
```

Comment (charleskawczynski) [reply docs/src/bcreader.md:23]: Oh, not sure my suggestion worked, basically I was suggesting to remove prepending spaces

File: src/BCReader.jl [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+# Returns
+- BCFileInfo
+"""
+function bcfile_info_init(
+    FT,
+    bcfile_dir,
+    datafile_rll,
+    varname,
+    boundary_space;
+    comms_ctx = ClimaComms.SingletonCommsContext(),
```
Comment (charleskawczynski) [inline src/BCReader.jl:108]:
```suggestion
    comms_ctx,
```
I think that creating separate `comms_ctx` may lead to strange mpi errors, so I'd suggesting making this a required argument
